File: python/perfect-numbers/perfect_numbers.py
from math import sqrt


def classify(number):
    if number <= 0:
        raise ValueError('Not a natural number')
    aliquot_sum = sum(f for f in range(1, number//2+1) if number % f == 0)
    if aliquot_sum == number:
        return 'perfect'
    elif aliquot_sum > number:
        return 'abundant'
    else:
        return 'deficient'

File: typescript/perfect-numbers/perfect-numbers.ts
enum NicomachusCategory {
  Perfect = 'perfect',
  Abudant = 'abundant',
  Deficient = 'deficient',
}

export default class PerfectNumbers {
  static classify(number: number): NicomachusCategory {
    if (number <= 0) {
      throw new Error('Classification is only possible for natural numbers.');
    }

    const factorsSum = PerfectNumbers.factorsSum(number);
    if (factorsSum == number) {
      return NicomachusCategory.Perfect;
    } else if (factorsSum > number) {
      return NicomachusCategory.Abudant;
    } else {
      return NicomachusCategory.Deficient;
    }
  }

  private static factorsSum(number: number): number {
    let sum = 0;
    for (let i = 1; i < number; i++) {
      if (number % i == 0) {
        sum += i;
      }
    }
    return sum;
  }
}
